File: algorithms/Linked List Cycle II/Linked List Cycle II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return None
        slow = fast = entry = head
        
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            
            if slow == fast:
                while slow != entry:
                    slow = slow.next
                    entry = entry.next
                return entry
        return None

File: algorithms/L/Linked List Cycle II/Linked List Cycle II.cpp
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        map<ListNode*, int> m;
        int index = 0;
        while (head)
        {
            if (m.find(head) == m.end())
                m[head] = index;
            else 
                return (head);
            head = head->next;
            index++;       
        }
        return (NULL);
    }
};
